File: 160.collections/src/main/scala/solution2/MyList.scala
package solution2

object MyList extends App {
  assert(args.size >= 1, "Usage: MyList 1 2 3 ...")

  val digits = args.map(_.toInt).toList
  val reversedDigits = reverse(digits)
  println(reversedDigits.mkString(" "))

  private def reverse(in: List[Int]): List[Int] = {
    if(in.isEmpty) List()
    else reverse(in.tail) :+ in.head
  }
}

File: 310.collections1/src/main/java/MyList.java
package javax;

import java.util.List;
import java.util.ArrayList;

public class MyList {
    public static void main(String[] args) {
        assert(args.length >= 1);

        List digits = new ArrayList();
        for(int i = 0; i < args.length; i++) {
            digits.add(Integer.parseInt(args[i]));
        }

        List reversedDigits = new ArrayList();
        reverse(digits, reversedDigits);
        for(int i = 0; i < reversedDigits.size(); i++) {
            System.out.print(reversedDigits.get(i) + " ");
        }
        System.out.println();
    }

    private static void reverse(final List in, List out) {
        for(int i = 0; i < in.size(); i++) {
            out.add(0, in.get(i));
        }
    }
}
